### ISSUE TITLE
CRAN CHECK ERROR - cachem::cache_mem

### ISSUE THREAD

**Issue by Polkas** [r-lib/cachem#25]
In one of my CRAN packages I am using memoise which is based on cachem. More precisely I am using memoise with max_age option like, `FUN <- memoise::memoise(FUN_RAW, cache = cachem::cache_mem(max_age = 60 * 60))`.
Unfortunately I get [this error](https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-fedora-gcc/pacs-00check.html) on R CRAN check (development R version), which looks to come from [that cachem function](https://github.com/r-lib/cachem/blob/f5ea146ca79eeb1dc806b1ba7039ff867e5b4bca/R/cache-mem.R#L229). The problem occurs only on linux fedora and R devel at this moment. Could it be some servers problems?
I will be glad to get some hints as have to solve it in a 1.5 week, not want to remove memoise functionality.
Thanks

**Comment by wch** [comment]
That is very strange. With the failing test, it looks like an `NA` is being passed to the `if` statement:

```
  > test_check("pacs")
  ══ Failed tests ════════════════════════════════════════════════════════════════
  ── Error (test-base.R:119:7): pacs::lib_validate ───────────────────────────────
  Error in `if (time - mtime_[idx] > max_age_) {
      log_(paste0("pruning single object exceeding max_age: Removing ",
          key))
      remove_(key)
      missing <- as_quosure(missing)
      return(eval_tidy(missing))
  }`: missing value where TRUE/FALSE needed
  Backtrace:
      █
   1. ├─pacs::lib_validate(...) test-base.R:119:6
   2. │ └─base::stopifnot(...)
   3. └─pacs::cran_flavors()
   4. └─pacs:::read_cran_flavours()
   5. └─encl$`_cache`$get(key)
  
  [ FAIL 1 | WARN 0 | SKIP 0 | PASS 82 ]
  Error: Test failures
  Execution halted
```

I think it may be that `time` or `mtime_[idx]` in this `if` statement is `NA` (https://github.com/r-lib/cachem/blob/f5ea146/R/cache-mem.R#L228-L234):

```R
      time <- as.numeric(Sys.time())
      if (time - mtime_[idx] > max_age_) {
        log_(paste0("pruning single object exceeding max_age: Removing ", key))
        remove_(key)
        missing <- as_quosure(missing)
        return(eval_tidy(missing))
      }
```

For example, an `NA` in a similar `if` statement results in the same error message:

```R
if (NA - 1234 > 123) {
  "OK"
}
#> Error in if (NA - 1234 > 123) { : missing value where TRUE/FALSE needed
```

Looking at the platforms that pass and fail at https://cran.r-project.org/web/checks/check_results_pacs.html, it appears that tests run on R-devel r81315 and earlier pass; tests run on r81321 and later fail. So it's possible that there was a bug introduced in R-devel somewhere between r81316 and r81321.

**Comment by Polkas** [comment]
Thank you for a quick response. I investigate if further and find out that it is a one code line problem, https://github.com/Polkas/pacs/blob/fd1a87948f8d1c79d379bd18d60ea8198f414f07/R/valid.R#L56. Looks like a usage of a memoised function in package compilation stage cause this error. When later such memoised function is used in the R session and other function which have it in its internals, the problem raise.  I think the usage time is started when package was build and some metadata is persist on continuously running server.
We have new R devel win builder https://win-builder.r-project.org/upload.aspx on which I check that the problem is only with this pointed line in my code. What is surprising, I have many similar cases in my code which not cause any problems. I see 4.2.0 might bring some fun:)

**Comment by wch** [comment]
After looking into this a bit more, I think that there's some sort of bug in R-devel, with the saving and restoring of objects in a package. Here's a way to reproduce using the `wch1/r-debug` Docker image:

```
docker run --rm -ti wch1/r-debug
RD
install.packages('pacs')

library(pacs)
# Get the memoise::cache_mem() object
cm <- environment(pacs:::read_cran_flavours)$`_cache`
cm$set("abc", 123)
cm$get("abc")
#> Error in if (time - mtime_[idx] > max_age_) { : 
#>   missing value where TRUE/FALSE needed
```

However, if pacs is loaded with `devtools::load_all()`, then there's no error.

This also doesn't happen when I tested on R 4.1.2 on my Mac.

I looked at the `time`, and there didn't appear to be a problem there. I think the problem has to something do with closures not correctly capturing the environment when they're saved in a package and then the package is loaded. I believe this is a bug in R-devel, although I'm not 100% sure yet.

**Comment by wch** [comment]
It's very strange that in my example above (with `read_cran_flavours`), it results in an error, but with `read_bio_releases` and `read_checkred_packages`, it does not error. 

```
library(pacs)
cm <- environment(pacs:::read_cran_flavours)$`_cache`
cm$set("abc", 123)
cm$get("abc")
#> Error in if (time - mtime_[idx] > max_age_) { : 
#>   missing value where TRUE/FALSE needed

cm <- environment(pacs:::read_bio_releases)$`_cache`
cm$set("abc", 123)
cm$get("abc")
#> [1] 123

cm <- environment(pacs:::read_checkred_packages)$`_cache`
cm$set("abc", 123)
cm$get("abc")
#> [1] 123
```

I'm almost certain that this is caused by a bug in R-devel.

**Comment by wch** [comment]
I'm no longer able to reproduce the error with R-devel (2021-12-14 r81376) and pacs 0.3.5 from CRAN, using the `wch1/r-debug` Docker image.

I suspect that the bug was fixed in R-devel by this commit (r81356):
 https://github.com/wch/r-source/commit/8cc8d1e142cc9c97d65bcfeb0e15fd36bbf56d8a

That commit also says the bug was introduced in r81320, which aligns with my observation above, which was:
>  tests run on R-devel r81315 and earlier pass; tests run on r81321 and later fail.

In short, I would expect the tests to fail on R-devel r81320 to r81356, due to the bug in R-devel. Versions outside of that range should be OK. That agrees with the test results I'm currently seeing on https://cran.r-project.org/web/checks/check_results_pacs.html.

If you find that the error still happens, you can reopen this issue.
